### ISSUE TITLE
Fix CSS-in-JS problems

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#4574]
We have a bunch of CSS-in-JS related problems. I've gathered them here:

- [ ] [#4119 — Fix incorrect output with autofixing for nested tagged template literals](https://github.com/stylelint/stylelint/issues/4119 )
- [ ] [#4140 — Fix false positives for array literal expressions](https://github.com/stylelint/stylelint/issues/4140)
- [ ] [#4490 — Fix autofix erroneous changes for float property](https://github.com/stylelint/stylelint/issues/4490)
- [ ] [#4560 — Fix autofix transforming media queries when no errors](https://github.com/stylelint/stylelint/issues/4560)
- [ ] [#4401 — Fix 'cannot parse selector' error for nested HTML elements](https://github.com/stylelint/stylelint/issues/4401)
- [ ] [#4247 — Support Vue-emotion syntax](https://github.com/stylelint/stylelint/issues/4247)

We recently forked [the parser](https://www.npmjs.com/package/postcss-jsx) we use for CSS-in-JS [into the stylelint organisation](https://github.com/stylelint/postcss-css-in-js). We can fix all the above issues by patching and maintaining this fork.

These issues are, along with the [glob issues](https://github.com/stylelint/stylelint/issues/4521) and the [one for disable comments](https://github.com/stylelint/stylelint/issues/2643), our most pressing ones right now. They all prevent people from using stylelint completely.

If you can help us fix our fork of the parser, please add a comment below!

**Comment by jeddy3** [comment]
As of `stylelint@13.3.1` we now use [our fork of the CSS-in-JS parser](https://github.com/stylelint/postcss-css-in-js), making it easier for us to roll-out any fixes to the parser that address the [issues above](https://github.com/stylelint/stylelint/issues/4574#issue-557025395). If you can help us fix our fork, please dive in!

**Comment by abdonrd** [comment]
Probably related: https://github.com/stylelint/stylelint/issues/4800

**Comment by rowa-audil** [comment]
Hey `stylelint` Team, any plans for #4713?

**Comment by Slapbox** [comment]
Is there any work ongoing on the forked parser? (https://github.com/stylelint/postcss-css-in-js)

**Comment by hudochenkov** [comment]
@Slapbox non that we aware of. In the next major version [v14](https://github.com/stylelint/stylelint/issues/5205) we're going remove it from stylelint.

I started working on another CSS-in-JS syntax, which would be possible to use with stylelint.

**Comment by jeddy3** [comment]
For context, the issues with the existing syntax are:
- it relies on [`postcss-syntax`](https://github.com/stylelint/postcss-syntax), which monkey-patches PostCSS@7
- it's monolithic; trying to support every CSS-in-JS flavour under the sun, of which there are hundreds ([list of just the 10 newest/most popular](https://github.com/andreipfeiffer/css-in-js/blob/main/README.md#overview))
- the original author is no longer active on GitHub

[Thanks to @hudochenkov](https://github.com/postcss/postcss/issues/1498), we now have a [`Document` node type in PostCSS@8](https://github.com/postcss/postcss/releases/tag/8.3.0).

Summary [of the plan for `14.0.0`](https://github.com/stylelint/stylelint/issues/5289#issuecomment-840772900):

- remove the `syntax` option in favour of `customSyntax`
- add an `overrides` property to the configuration object to aid the use of `customSyntax`
- encourage the community to [write custom PostCSS syntaxes](https://github.com/postcss/postcss/blob/main/docs/syntax.md) that focus on a single CSS-in-JS library and make use of the `Document` node type

A [lot has already been done for `14.0.0`](https://github.com/stylelint/stylelint/blob/v14/CHANGELOG.md#head), including a [a chunk of refactoring](https://github.com/stylelint/stylelint/pulls?q=is%3Apr+is%3Aclosed+refactor). Importantly, the [`v14` branch](https://github.com/stylelint/stylelint/tree/v14) has already been migrated to PostCSS@8, and the `customSyntax` option is ready to use as a CLI flag.

@Slapbox If you, or anyone else, would like to help us get `14.0.0` ready for release, please help [with the remaining tasks](https://github.com/stylelint/stylelint/issues/5205). In particular:

- adding the `customSyntax` and `overrides` properties to the configuration object
- migrating [`postcss-html`](https://github.com/stylelint/postcss-html) and [`postcss-markdown`](https://github.com/stylelint/postcss-markdown) to PostCSS@8 (and the new `Document` node type)

Once @hudochenkov reference syntax is ready, we can start encouraging the community to work on syntaxes for their library of choice.

**Comment by heath-freenome** [comment]
> @Slapbox non that we aware of. In the next major version [v14](https://github.com/stylelint/stylelint/issues/5205) we're going remove it from stylelint.
> 
> I started working on another CSS-in-JS syntax, which would be possible to use with stylelint.

So I'm migrating to stylelint 14 and all my styles are in .js files (using the Material-ui 4 JSS syntax). Using the `@stylelint/postcss-css-in-js` doesn't seem to pick up any of the rules I have for my styles... will yours
 work?

**Comment by Lornz-** [comment]
> > @Slapbox non that we aware of. In the next major version [v14](https://github.com/stylelint/stylelint/issues/5205) we're going remove it from stylelint.
> > I started working on another CSS-in-JS syntax, which would be possible to use with stylelint.
> 
> @hudochenkov So I'm migrating to stylelint 14 and all my styles are in .js files (using the Material-ui 4 JSS syntax). Using the customSyntax `@stylelint/postcss-css-in-js` doesn't seem to pick up any of the rules I have for my styles... will yours work?

Same issue here.I **only get it worked from the command line**. I use **styled-components** in one of my projects and the code is not highlighted on error in vscode. It seems like the vscode extension can't see the rules in my config file. I also combine stylelint with prettier by using the **stylelint-prettier** plugin. 

I followed steps as described [here](https://github.com/styled-components/styled-components/issues/3607#issuecomment-983637806) with my own setup:

**package.json**
```json
{
  "scripts": {
    // other scripts omitted for brevity
   "lint:style": "npx stylelint \"**/*.{css,jsx,js}\"",
  },
  "dependencies": {
    // other deps omitted for brevity
    "styled-components": "^5.3.0",
  },
  "devDependencies": {
    // other deps omitted for brevity
    "@stylelint/postcss-css-in-js": "^0.37.2",
    "postcss": "^8.4.5",
    "postcss-syntax": "^0.36.2",
    "stylelint": "^14.2.0",
    "stylelint-a11y": "^1.2.3",
    "stylelint-config-prettier": "^9.0.3",
    "stylelint-config-rational-order": "^0.1.2",
    "stylelint-config-recommended": "^6.0.0",
    "stylelint-config-standard": "^24.0.0",
    "stylelint-config-styled-components": "^0.1.1",
    "stylelint-high-performance-animation": "^1.5.1",
    "stylelint-no-unsupported-browser-features": "^5.0.2",
    "stylelint-order": "^5.0.0",
    "stylelint-prettier": "^2.0.0",
  },
}
```

**.stylelintrc.js**
```js
module.exports = {
  plugins: [
    'stylelint-a11y',
    'stylelint-config-rational-order/plugin',
    'stylelint-high-performance-animation',
    'stylelint-no-unsupported-browser-features',
    'stylelint-order',
    'stylelint-prettier',
  ],
  extends: [
    'stylelint-config-standard',
    'stylelint-config-recommended',
    'stylelint-config-rational-order',
    'stylelint-a11y/recommended',
    'stylelint-prettier/recommended',
    'stylelint-config-styled-components',
  ],
  overrides: [
    {
      files: ['**/*.{js,jsx}'],
      customSyntax: '@stylelint/postcss-css-in-js',
    },
  ],
```

Here's also my vscode settings for the matter:
```json
{
  // other settings omitted for brevity
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "eslint.format.enable": true,
  "editor.formatOnSave": true,
  "editor.codeActionsOnSave": {
    "source.formatDocument": true,
    "source.fixAll": true
  },
}
```

Did someone find a solution for that? Any help would be appreciated.

**Comment by puku0x** [comment]
@Lornz-  I did.

```js
// package.json
"devDependencies": {
  "@babel/core": "7.16.0",  // Do not update babel v7.16.5 or higher (https://github.com/stylelint/postcss-css-in-js/issues/247)
  "@babel/parser": "7.16.0",
  "postcss-jsx": "~0.36.0",  // Do not use stylelint/postcss-css-in-js with decorators (https://github.com/stylelint/postcss-css-in-js/issues/236)
  "postcss-scss": "~4.0.0",
  "postcss-syntax": "~0.36.0",
  "stylelint": "~14.2.0",
  "stylelint-config-prettier": "~9.0.0",
  "stylelint-config-recess-order": "~3.0.0",
  "stylelint-config-recommended": "~6.0.0"
}
```

```js
// .stylelintrc.json
{
  "extends": [
    "stylelint-config-recommended",
    "stylelint-config-recess-order",
    "stylelint-config-prettier"
  ],
  "ignoreFiles": ["node_modules/**", "dist/**"],
  "overrides": [
    {
      "files": ["**/*.scss"],
      "customSyntax": "postcss-scss"
    },
    {
      "files": ["**/*.js", "**/*.jsx", "**/*.ts", "**/*.tsx"],
      "customSyntax": "postcss-jsx"
    }
  ]
}
```
```js
// .vscode/settings.json
{
  "css.validate": false,
  "scss.validate": false,
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true,
    "source.fixAll.stylelint": true
  },
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "editor.formatOnSave": true,
  "stylelint.validate": ["css", "scss", "typescript", "typescriptreact"]  // 👈 Use stylelint for validation
}
```
